lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1407,25 +1425,123 @@ static void diagnoseImplicitSelfUseInClosure(TypeChecker &TC, const Expr *E,
+        }
+        
+        return true;
+      }
+      return false;
+    }
+    
+    /// Emit fix-its for invalid use of implicit \c self in an explicit closure.
+    /// The error can be solved by capturing self explicitly,
+    /// or by using \c self. explicitly.
+    void emitFixItsForExplicitClosure(SourceLoc memberLoc,
+                                      const ClosureExpr *closureExpr) {
+      TC.diagnose(memberLoc, diag::note_reference_self_explicitly)
+        .fixItInsert(memberLoc, "self.");
+      auto diag = TC.diagnose(closureExpr->getLoc(),
+                              diag::note_capture_self_explicitly);
+      // There are four different potential fix-its to offer based on the
+      // closure signature:
+      //   1. The signature empty so far. We must insert the full capture
+      //      list as well as 'in'.
+      //   2. Arguments or types are already specified in the signature,
+      //      but there is no existing capture list. We will need to insert
+      //      the capture list, but 'in' will already be present.
+      //   3. There is an existing capture list which already has some
+      //      entries. We need to insert 'self' into the capture list along
+      //      with a separating comma.
+      //   4. There is an existing capture list, but it is empty. We can
+      //      just insert 'self'.
+      auto brackets = closureExpr->getBracketRange();
+      if (brackets.isInvalid()) {
+        if (closureExpr->getInLoc().isInvalid()) {
+          // If there's a (non-comment) token immediately following the
+          // opening brace of the closure, we may need to pad the fix-it
+          // with a space.
+          const auto nextLoc = closureExpr->getLoc().getAdvancedLoc(1);
+          const auto next =
+              Lexer::getTokenAtLocation(TC.Context.SourceMgr, nextLoc,
+                                        CommentRetentionMode::None);
+          std::string trailing = next.getLoc() == nextLoc ? " " : "";
+
+          diag.fixItInsertAfter(closureExpr->getLoc(), " [self] in" + trailing);
+        }
+        else
+          diag.fixItInsertAfter(closureExpr->getLoc(), " [self]");
+      }
+      else {
+        // Similar to above, we look for any non-comment token. If there's
+        // anything before the closing bracket, we assume that it is a valid
+        // capture list entry and insert 'self,'. If it wasn't a valid
+        // entry, then we will at least not be introducing any new
+        // errors/warnings...
+        const auto locAfterBracket = brackets.Start.getAdvancedLoc(1);
+        const auto nextAfterBracket =
+            Lexer::getTokenAtLocation(TC.Context.SourceMgr, locAfterBracket,
+                                      CommentRetentionMode::None);
+        if (nextAfterBracket.getLoc() != brackets.End)
+          diag.fixItInsertAfter(brackets.Start, "self, ");
+        else
+          diag.fixItInsertAfter(brackets.Start, "self");
+      }
+    }
```

Comment:
I think that the control flow can be made clearer by changing 1501 to `brackets.isValid()`. Then you can have the existing else clause as the then clause with an early return. Then `if (closureExpr->getInLoc().isValid()` with just the one line (from 1515) and another early return. And finish up with the existing 1503-1512.

The principle is to choose the logical parity of if predicates to put the simpler case first. That reduces the depth of the mental stack of the reader. The early return can be a mixed bag, but here since it is unlikely that cleanup will be added, it seems like a better choce than an else.

Reply:
Yep, makes sense. Will do!

Reply:
Thanks!

Reply:
I went ahead and pulled out each branch of the `brackets.isValid()` check into a separate function as well. Once tests are passing locally I'll put up the latest changes!